src/Storages/StorageLogSettings.cpp

```diff
@@ -1,18 +1,42 @@
 #include "StorageLogSettings.h"
+#include <Disks/StoragePolicy.h>
+#include <Interpreters/Context.h>
 #include <Parsers/ASTCreateQuery.h>
 #include <Parsers/ASTSetQuery.h>
 
 namespace DB
 {
-String getDiskName(ASTStorage & storage_def)
+
+namespace ErrorCodes
+{
+    extern const int INVALID_SETTING_VALUE;
+}
+
+String getDiskName(ASTStorage & storage_def, ContextPtr context)
 {
     if (storage_def.settings)
     {
         SettingsChanges changes = storage_def.settings->changes;
-        for (const auto & change : changes)
-            if (change.name == "disk")
-                return change.value.safeGet<String>();
+
+        const auto disk_change
+            = std::find_if(changes.begin(), changes.end(), [&](const SettingChange & change) { return change.name == "disk"; });
+        const auto storage_policy_change
+            = std::find_if(changes.begin(), changes.end(), [&](const SettingChange & change) { return change.name == "storage_policy"; });
```

Comment:
this could be done in 1 for loop, now it adds some complexity, but ok, the list of settings is usually not big